File: common/src/main/java/dev/number6/slack/port/SlackPort.java
package dev.number6.slack.port;

import com.amazonaws.services.lambda.runtime.LambdaLogger;
import okhttp3.RequestBody;

import java.util.Optional;

public interface SlackPort {

    Optional<String> getSlackResponse(String url, LambdaLogger logger, String body);

    <T> Optional<T> getSlackResponse(String url, Class<T> responseType, LambdaLogger logger);

    <T> Optional<T> getSlackResponse(String url, String body, Class<T> responseType, LambdaLogger logger);
}

File: common/src/main/java/dev/number6/slack/port/SlackPort.kt
package dev.number6.slack.port

import com.amazonaws.services.lambda.runtime.LambdaLogger
import java.util.*

interface SlackPort {
    fun getSlackResponse(url: String, logger: LambdaLogger, body: String): Optional<String>
    fun <T> getSlackResponse(url: String, responseType: Class<T>?, logger: LambdaLogger): Optional<T>
    fun <T> getSlackResponse(url: String, body: String?, responseType: Class<T>?, logger: LambdaLogger): Optional<T>
}